### DISCUSSION TITLE
Our experiences with Shadow when simulating BFT protocols

### DISCUSSION THREAD

**Discussion by bergerch** [shadow/shadow#2636] [Show and tell]
The latest blockchain generation of Byzantine fault-tolerant (BFT) state machine replication (SMR) protocols focuses on scalability and performance to meet the requirements of distributed ledger technology (DLT), such as decentralization and geographic dispersion. Validating the scalability and performance of BFT protocol implementations requires careful evaluation. 

In this field, experiments with actual protocol deployments offer the best realism, but they can be costly and time-consuming. We investigated whether simulation can be a fitting (i.e., resource-friendly but still high-fidelity) alternative to study the performance and scalability of BFT protocols. 

For this purpose, we explored the performance (through benchmarks) of various BFT protocol implementations using the Shadow simulator, in particular we chose the well-known protocols PBFT, BFT-SMaRt, and HotStuff, with their implementations written in Rust, Java, and C++, respectively.

A big motivator for using Shadow was preserving application realism, as existing BFT frameworks can be easily plugged into the simulation engine without requiring code modifications or re-implementation. Note that these BFT protocols tend to be extremely complex, thus making a re-implementation or creating an application model very error-prone and unreliable. Such attempts would require justification about whether the model still sufficiently matches the actual protocol.

Our experiments mostly centered around two questions:

1. 	 Accuracy. How close will simulation results approximate the results of real protocol runs?
2. 	 Resource Usage: How resource-friendly are such simulations in terms of memory/time

Since Shadow is a network simulator and does not capture CPU intensive tasks such as creating or validating cryptographic message authenticators (which can be a performance limiting factor in BFT protocols), we were especially interested in how the accuracy might vary for different system setups, in particular when scaling the system up to many replicas.  

I want to briefly preempt some of our main insights:
-  In larger scale systems (many replicas), the network bandwidth becomes a limiting factor for system performance. In this case, Shadow simulations can provide high-fidelity results.
-  In experiments that focus on measure only latency, and use few workload, (i.e., the workload does not saturate the system), the latency prediction is also very good.
-  In experiments with a small number of replicas or when requests do not carry any payload, simulations will overestimate system throughput. This is because in real executions, the CPU of the replicas (in particular their capability of validating signatures) becomes the performance limiting factor.

<img src="https://user-images.githubusercontent.com/5364442/210340296-fcaf4c4e-ac0e-4c91-b944-036f082f1073.svg" width="505" height="230">

In one of our experiments, we sought to compare the results of our simulation with the measurements presented in the HotStuff paper by replicating their evaluation setup. This setup included the use of over 100 virtual machines located in an AWS data center, each with a maximum bandwidth of 1.2 GB/s. The latency between any pair of machines was less than 1 ms (we used 1 ms in our simulation). The HotStuff system used a block size of 400 transactions per block, with each transaction having a payload of 1024 bytes.

The simulation results for the payload experiment show a trend similar to the real measurements, where performance begins to decline for n ≥ 32. For a small replica group, the network simulation predicts higher performance, reaching a maximum of 200k tx/s. This is limited only by the 1 ms link latency, allowing HotStuff to commit a batch of 400 requests every 2 ms. As the performance of HotStuff becomes more constrained by bandwidth (for n ≥ 32), the difference in observed throughput decreases.

<img src="https://user-images.githubusercontent.com/5364442/210340370-f99dbbd8-7ce4-4d19-9e25-cd5d6c9d4c38.svg" width="505" height="184">

When simulating different BFT protocol implementations
-  BFTSMaRt (Java)
-  HotStuff (C++)
-  PBFT/Themis (Rust)

with 1KiB request and response payloads, and increased the system scale, our observations showed that PBFT initially performs better than HotStuff, but its throughput decreases more rapidly (as seen in the steeper curve). At n = 128, PBFT reaches a throughput of up to 9.3k tx/s, while HotStuff achieves a throughput of up to 20k tx/s. BFT-SMaRt, which uses the same agreement pattern as PBFT, also reaches a throughput of up to 9.2k tx/s at similar latency values.

![resourceConsumption](https://user-images.githubusercontent.com/5364442/210340495-7f1dfd3f-11b4-4a4d-b5e0-abd6e687af8a.svg)

**Memory Usage:** BFT-SMaRt experiments reported the highest memory usage, increasing linearly, in spite of the protocol being under a much lower fixed load, using up 135 GB at 𝑛 = 128 replicas compared to PBFT’s 64 GB and HotStuff’s 42 GB. This is most likely due to the need for the Java Runtime to be loaded, which is not the case for C++ and Rust programs.

**Benchmark Time:** When simulating a total of  210 seconds, 150 of which were reserved for bootstrapping and the remaining 60 seconds were benchmark time,  The duration increased linearly with the system scale for BFT-SMaRt whereas for HotStuff and PBFT it was somewhat constant. In particular, at n=128 BFT-SMaRt simulations took considerably longer with 405 minutes compared to 150 minutes for PBFT and 78 minutes for HotStuff.

**Hiccups:**

- We noticed that Shadow fails to kill all java processes when simulating BFT-SMaRt (usually 1 process survives the kill)
- BFT-SMaRt simulations sometimes deadlock. This is possibly related to the Busy Looping issue. We unfortunately have not yet tried the –model-unblocked-syscall-latency workaround but we will in the near future.
- We encountered the following unsupported system calls:
	81 (fchdir): BFT-SmaRt
            98 (getrusage)
	204 (sched_getaffinity): Themis, BFT-SMaRt
            229 (clock_getres): libhotstuff (HotStuff), BFT-SMaRt

To bootstrap our simulastions, we built a tool that acts as a frontend towards Shadow and thus simplifies and accelerate the evaluation of unmodified BFT protocol implementations. It can orchestrate several simulations, generate realistic network environments based on the AWS latency map, configure and build several BFT protocols, create runtime artifacts and monitor resource usage.  For interested readers: we described the architecture of this tool and our evaluations with BFT protocols in a [paper](https://dicg-workshop.github.io/2022/papers/berger.pdf). The tool is open-source available here: https://github.com/Delphi-BFT/tool

**Comment by robgjansen** [comment]
Thanks for the detailed write-up! We are excited to see Shadow being used for research and evaluation of interesting applications, and it seems to be a great fit for your work :)

And thanks for reporting the hiccups, those should help us make some improvements to Shadow for future work.

WRT the observation that since Shadow does not yet model CPU processing overhead, it tends to overestimate throughout when CPU is the limiting factor: we are curious if "simulating" the CPU load could give more accurate results in that case. This could be done, for example, by adding in some calls to `nanosleep()` from the application to force the application to block for an amount of time equal to the expected CPU processing time for the various crypto operations. Did you try this and did it have any effect on Shadow's throughout accuracy?

**Reply by bentoumia** [reply]
Thank you for the suggestion @robgjansen! Unfortunately, we have not tried this yet, but we might in the near future :)

**Comment by stevenengler** [comment]
> We noticed that Shadow fails to kill all java processes when simulating BFT-SMaRt (usually 1 process survives the kill)

Do you have a minimal example that can reproduce this? Or any comments/notes that would help us reproduce it? This is definitely something we'd want to look into if we're able to.
